WalletWasabi/Services/WalletService.cs

```diff
@@ -904,24 +896,27 @@ public async Task<int> CountBlocksAsync()
 			}
 
 			// Get allowed coins to spend.
-			List<SmartCoin> allowedSmartCoinInputs; // Inputs that can be used to build the transaction.
+			var coinsView = Coins.Available();
+			List<SmartCoin> allowedSmartCoinInputs = allowUnconfirmed // Inputs that can be used to build the transaction.
+				? coinsView.ToList()
+				: coinsView.Confirmed().ToList();
+
 			if (allowedInputs != null) // If allowedInputs are specified then select the coins from them.
 			{
 				if (!allowedInputs.Any())
 				{
 					throw new ArgumentException($"{nameof(allowedInputs)} is not null, but empty.");
 				}
 
-				allowedSmartCoinInputs = allowUnconfirmed
-					? Coins.Where(x => !x.Unavailable && allowedInputs.Any(y => y.TransactionId == x.TransactionId && y.Index == x.Index)).ToList()
-					: Coins.Where(x => !x.Unavailable && x.Confirmed && allowedInputs.Any(y => y.TransactionId == x.TransactionId && y.Index == x.Index)).ToList();
+				var availableCoinsView = coinsView
+					.FilterBy(x => allowedInputs.Any(y => y.TransactionId == x.TransactionId && y.Index == x.Index));
 
 				// Add those that have the same script, because common ownership is already exposed.
 				// But only if the user didn't click the "max" button. In this case he'd send more money than what he'd think.
 				if (payments.ChangeStrategy != ChangeStrategy.AllRemainingCustom)
 				{
 					var allScripts = allowedSmartCoinInputs.Select(x => x.ScriptPubKey).ToHashSet();
-					foreach (var coin in Coins.Where(x => !x.Unavailable && !allowedSmartCoinInputs.Any(y => x.TransactionId == y.TransactionId && x.Index == y.Index)))
+					foreach (var coin in coinsView.FilterBy(x => !allowedSmartCoinInputs.Any(y => x.TransactionId == y.TransactionId && x.Index == y.Index)))
```

Comment:
As I mentioned at the PR, there is a bug here somewhere. Load up a RegTest wallet. Go to SentTab. Select one Coin. Press Max button. Send the tx and all of your coins will be selected as an input to build the transaction. 
First please write a failing test into a **separate** PR to detect this.

Reply:
Good catch. I made a mistake during the refactoring. No need to test for reproducing it because the error was evident but I will create a test suite for the transaction building method anyway.

Reply:
Now we have a decent number of unit tests that guarantee bugs like this one don't go unnoticed.